src/ripple/protocol/TER.h

```diff
@@ -218,6 +292,140 @@ enum TER
     tecEXPIRED                  = 148
 };
 
+// Allow TECcodes to be promoted to TER.
+template <>
+class CanCvtToTER<TECcodes> : public std::true_type
+{ explicit CanCvtToTER() = delete; };
+
+//------------------------------------------------------------------------------
+
+// Template class that is specific to selected ranges of error codes.  The
+// Trait tells std::enable_if which ranges are allowed.
+template <template<typename> class Trait>
+class TERSubset
+{
+    TERUnderlyingType code_;
+
+public:
+    // Constructors
+    constexpr TERSubset() : code_ (tesSUCCESS) { }
+    constexpr TERSubset (TERSubset const& rhs) = default;
+    constexpr TERSubset (TERSubset&& rhs) = default;
+
+private:
+    constexpr TERSubset (int rhs) : code_ (rhs) { } // For TxMeta
+public:
+    // A factory function.  This method must be called to construct from int.
+    constexpr static TERSubset fromInt (int value) { return TERSubset(value); }
+
+    template <typename T, typename = std::enable_if_t<Trait<T>::value>>
+    constexpr TERSubset (T rhs) : code_ (static_cast<int>(rhs)) { }
+
+    // Assignment
+    constexpr TERSubset& operator=(TERSubset const& rhs) = default;
+    constexpr TERSubset& operator=(TERSubset&& rhs) = default;
+
+    template <typename T>
+    constexpr auto
+    operator= (T rhs) -> std::enable_if_t<Trait<T>::value, TERSubset&>
+    { code_ = static_cast<int>(rhs); return *this; }
+
+    // Conversion operators.
+    // Conversion to int is not explicit to allow streaming without casting.
+    constexpr operator int() const { return code_; }
+
+    // Conversion to Json::Value allows assignment to Json::Objects
+    // without casting.
+    operator Json::Value() const { return Json::Value {code_}; }
+
+    // Comparison
+    friend constexpr bool operator== (TERSubset lhs, TERSubset rhs)
+    { return lhs.code_ == rhs.code_; }
+
+    friend constexpr bool operator!= (TERSubset lhs, TERSubset rhs)
+    { return ! (lhs == rhs); }
```

Comment:
We need to write three implementations of the comparison functions for each comparison: one for `template<template<typename> class T1, template<typename> class T2>`, one for `TERSubset, enum type`, and one for `enum type, TERSubset`. It would be a win to get this down to one function. Maybe a `detail` scoped free function that converts both the enums and `TERSubset` to an int, do a `enable_if` based on if that free function is callable, then write the one comparison function.